src/main/kotlin/uk/gov/justice/digital/hmpps/prisonregister/service/PrisonService.kt

```diff
@@ -255,4 +222,126 @@ class PrisonService(
+  }
+
+  @Transactional
+  @PreAuthorize(CLIENT_CAN_MAINTAIN_ADDRESSES)
+  fun updateContactDetails(prisonId: String, updateContactDetailsDto: ContactDetailsDto, removeIfNull: Boolean = true): ContactDetailsDto {
+    prisonRepository.getReferenceById(prisonId) ?: throw EntityNotFoundException()
+
+    val contactDetails =
+      contactDetailsRepository.getByPrisonIdAndType(prisonId, updateContactDetailsDto.type) ?: throw ContactDetailsNotFoundException(
+        prisonId,
+        updateContactDetailsDto.type,
+      )
+
+    if (haveContactDetailsChanged(updateContactDetailsDto, contactDetails)) {
+      val originalDetails = ContactDetailsDto(contactDetails)
+
+      updateContactDetailsDto.emailAddress?.let {
+        val emailAddress = createOrGetEmailAddress(it)
+        emailAddress.contactDetails.add(contactDetails)
+        contactDetails.emailAddress = emailAddress
+      } ?: run { if (removeIfNull) contactDetails.emailAddress = null }
+
+      updateContactDetailsDto.webAddress?.let {
+        val webAddress = createOrGetWebAddress(it)
+        webAddress.contactDetails.add(contactDetails)
+        contactDetails.webAddress = webAddress
+      } ?: run { if (removeIfNull) contactDetails.webAddress = null }
+
+      updateContactDetailsDto.phoneNumber?.let {
+        val phoneNumber = createOrGetPhoneNumber(it)
+        phoneNumber.contactDetails.add(contactDetails)
+        contactDetails.phoneNumber = phoneNumber
+      } ?: run { if (removeIfNull) contactDetails.phoneNumber = null }
+
+      val persistedEntity = contactDetailsRepository.saveAndFlush(contactDetails)
+
+      removeOrphanedContactDetails(originalDetails)
+
+      return ContactDetailsDto(persistedEntity)
+    }
+
+    return ContactDetailsDto(contactDetails)
+  }
+
+  @Transactional
+  @PreAuthorize(CLIENT_CAN_MAINTAIN_ADDRESSES)
+  fun deleteContactDetails(prisonId: String, type: DepartmentType) {
+    val contactDetails =
+      contactDetailsRepository.getByPrisonIdAndType(prisonId, type) ?: throw ContactDetailsNotFoundException(
+        prisonId,
+        type,
+      )
+
+    contactDetailsRepository.delete(contactDetails)
+
+    removeOrphanedContactDetails(ContactDetailsDto(contactDetails))
+  }
+
+  private fun removeOrphanedContactDetails(contactDetails: ContactDetailsDto) {
+    contactDetails.emailAddress?.let {
+      if (contactDetailsRepository.isEmailOrphaned(it)) {
+        emailAddressRepository.delete(it)
+      }
+    }
+
+    contactDetails.phoneNumber?.let {
+      if (contactDetailsRepository.isPhoneNumberOrphaned(it)) {
+        phoneNumberRepository.delete(it)
+      }
+    }
+
+    contactDetails.webAddress?.let {
+      if (contactDetailsRepository.isWebAddressOrphaned(it)) {
+        webAddressRepository.delete(it)
+      }
+    }
+  }
+
+  private fun haveContactDetailsChanged(updateContactDetailsDto: ContactDetailsDto, contactDetails: ContactDetails): Boolean {
+    return updateContactDetailsDto.type != contactDetails.type ||
```

Comment:
.equals after converting to ContactDto?

Reply:
DTO should not be used to compare two objects that are not necessarily the same.